src/redisearch_rs/result_processor/src/lib.rs

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (c) 2006-Present, Redis Ltd.
+ * All rights reserved.
+ *
+ * Licensed under your choice of the Redis Source Available License 2.0
+ * (RSALv2); or (b) the Server Side Public License v1 (SSPLv1); or (c) the
+ * GNU Affero General Public License v3 (AGPLv3).
+*/
+
+pub mod counter;
+pub mod ffi;
+
+use std::{ffi::c_void, marker::PhantomData, mem::MaybeUninit, ptr::NonNull};
+
+pub enum Error {
+    // stub
+}
+
+#[repr(i32)]
+#[non_exhaustive]
+pub enum ResultProcessorType {
+    Index,
+    Loader,
+    SafeLoader,
+    Scorer,
+    Sorter,
+    Counter,
+    PageLimiter,
+    Highlighter,
+    Group,
+    Projector,
+    Filter,
+    Profile,
+    Network,
+    Metrics,
+    KeyNameLoader,
+    MaxScoreNormalizer,
+    Timeout, // DEBUG ONLY
+    Crash,   // DEBUG ONLY
+    Max,
+}
+
+/// This is the main trait that Rust result processors need to implement
+pub trait ResultProcessor {
+    const TYPE: ResultProcessorType;
+
+    fn next(&mut self, cx: Context) -> Result<Option<ffi::SearchResult>, Error>;
+}
+
+/// This type allows result processors to access its context (the owning QueryIterator, upstream result processors, etc.)
+pub struct Context<'a> {
+    rp: &'a mut ffi::Header,
+}
+
+impl Context<'_> {
+    /// The QueryIterator that owns this result processor
+    pub fn parent(&mut self) -> NonNull<c_void> {
+        todo!()
+    }
+
+    /// The previous result processor in the pipeline
+    pub fn upstream(&mut self) -> Option<Upstream<'_>> {
+        let rp = self.rp.upstream?;
+
+        Some(Upstream {
+            rp,
+            _m: PhantomData,
+        })
+    }
+}
+
+/// The previous result processor in the pipeline
+pub struct Upstream<'a> {
+    rp: NonNull<ffi::Header>,
```

Comment:
nit: I'd suggest more descriptive names (unless it's a temporary variable used in a very small scope). `result`, `r_processor` or even `result_processor` would be good IMO.